### PR TITLE
Function for array permutation

### PR DESCRIPTION
**Pull request by mcorrell**
Permutes an array in place. Permuted with a Knuth shuffle aka the Fisher-Yates shuffle.

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -233,6 +233,21 @@ u.stablesort = function(array, sortBy, keyFn) {
...
 };
 
+// permutes an array using a Knuth shuffle
+u.permute = function(a) {
+  var j,
+      toSwap,
+      i = 0,
+      array = a.slice(0);
+  
+  for(i; i<a.length; ++i){
```
Comment (domoritz) [inline src/util.js]: How about

```js
  while (m) {
    i = Math.floor(Math.random() * m--);
    t = array[m];
    array[m] = array[i];
    array[i] = t;
  }
```

Totally taken from https://bost.ocks.org/mike/shuffle/

**Comment by mcorrell** [comment]
For my own code I prefer incrementing for loops over decrementing while loops. Both are O(n). The decrementing version I think will be slightly faster in practice (the space of possible swaps gets smaller and smaller with each decrement, for instance). The decrementing version is also technically more accurate, since the incrementing version allows multiple swaps of the same value (which means the identity permutation is slightly less likely, among other properties). 

Switching over is certainly easy enough.

**Comment by domoritz** [comment]
I don't care about the performance difference but found the decrementing version or be easier.

**Comment by domoritz** [comment]
More importantly, why do you create a copy when the explicit advantage of the algorithm is that you can shuffle in place? If it is explicit (e.g. no return value), users can make a copy themselves if they need to.

**Comment by mcorrell** [comment]
Also personal preference: I prefer these methods to not have side effects.

But precedent both externally (for instance php's std `shuffle()` method) and internally (most of the other array methods in util have side effects) backs you up, so I'll switch over.

**Comment by domoritz** [comment]
LGTM. I changed the tests a bit but this looks good to go.

**Pull request merged**